core/src/main/scala/com/spotify/featran/FeatureExtractor.scala

```diff
@@ -116,10 +116,27 @@ class RecordExtractor[T, F: FeatureBuilder: ClassTag] private[featran] (fs: Feat
       }
     }
 
-  private val input: PipeIterator = new PipeIterator
-  private val extractor: FeatureExtractor[Iterator, T] =
-    new FeatureExtractor[Iterator, T](fs, input, Some(Iterator.continually(settings)))
-  private val output: Iterator[FeatureResult[F, T]] = extractor.featureResults
+  private final case class State(private val input: PipeIterator,
+                                 private val extractor: FeatureExtractor[Iterator, T],
+                                 private val output: Iterator[FeatureResult[F, T]]) {
+    val featureNames: Seq[String] = extractor.featureNames.next()
+
+    def featureResult(record: T): FeatureResult[F, T] = {
+      input.feed(record)
+      output.next()
+    }
+  }
+
+  private val state = new ThreadLocal[State] {
+    override def initialValue(): State = {
+      val input: PipeIterator = new PipeIterator
+      val extractor: FeatureExtractor[Iterator, T] =
+        new FeatureExtractor[Iterator, T](fs, input, Some(Iterator.continually(settings)))
```

Comment:
This is the expensive part we're trying to avoid. This could still be a problem if we have a lot of short-lived threads (like in Dataflow streaming?). I wonder if there's a better way to solve this? Like making `FeatureExtractor` easily cloneable and share the parsed settings?